### PR TITLE
[GCP] Make sure the chosen VPC has subnet enabled for the region required

### PR DESCRIPTION
**Pull request by Michaelvll**
A user met an issue when SkyPilot selects a VPC without subnet enabled for the region required (us-central2 for TPU-v4). This PR enforce the provisioner to select the VPC that has the subnet enabled.

This PR also fixes a corner case, when the user without TPU-v4 quota is using `n1-highmem-8` and the failover goes to `us-central2`. This will fail before, as the user will not have subnet in `us-central2`

Tested (run the relevant ones):

- [x] Code formatting: `bash format.sh`
- [x] Any manual or new tests for this PR (please specify below)
  - [x] `sky tpunode --tpus tpu-v4-8 --tpu-vm` on project without us-central2 enabled. It should fail on the region and normally trying to failover to other region.
  - [x] `sky launch -c test --cloud gcp` on a project with 2 VPCs, one only with us-central1 subnet enabled, another with all subnets.
  - [x] `sky launch` an existing cluster.
- [ ] All smoke tests: `pytest tests/test_smoke.py` 
- [ ] Relevant individual smoke tests: `pytest tests/test_smoke.py::test_fill_in_the_name` 
- [ ] Backward compatibility tests: `bash tests/backward_comaptibility_tests.sh`

### CODE REVIEW

**Comment by Michaelvll** [comment]
Blocked by #2764

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:849]: How about changing get_usable_vpc() into get_usable_vpc_and_subnets(), i.e., avoid an extra query? 


https://github.com/skypilot-org/skypilot/blob/31aa76c968e6c9ca8c21ac25051f002383c1baee/sky/provision/aws/config.py#L72-L75

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:794]: In AWS, we get a usable VPC first, then get its usable subnets:

https://github.com/skypilot-org/skypilot/blob/31aa76c968e6c9ca8c21ac25051f002383c1baee/sky/provision/aws/config.py#L399-L414

Any pros/cons in terms of doing it like that vs. the nested loop here? Maybe the former makes it easier to support things like private IP-enabled subnets in the future ?

**Review by concretevitamin — COMMENTED** [review]
Thanks for the quick fix @Michaelvll. Quick comments.

File: sky/skylet/providers/gcp/config.py
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py:794]: Thanks for mentioning this @concretevitamin! We do need to sort the VPCs in this GCP case as well.

While looking at the code for AWS, it seems that we did not sort the subnets by the vpc id, which means when the user is creating multiple clusters without specifying a VPC name, it is possible that those clusters can locate in different VPCs. It might be fine for now as there is no interconnection between clusters for us right now, but it does not seem very clean. Should we sort the subnets by the (availability_zone, vpc_id) pair?

For GCP here, yes, we can do similar thing for getting the VPC. The current implementation was just inherited from the previous implementation from ray. I will start modifying it once #2764 is merged.

Comment (concretevitamin) [reply sky/skylet/providers/gcp/config.py:794]: Not too sure I understand the parts about sorting. Some questions:

> which means when the user is creating multiple clusters without specifying a VPC name, it is possible that those clusters can locate in different VPCs

This seems okay? E.g., users may change the VPC name setting in ~/.sky/config.yaml across launches, and SkyPilot does faithfully change the VPC to use.

Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py:794]: Oh, I mean If the user does not have `~/.sky/config.yaml`, it is possible that multiple clusters may be located on different VPCs depends on the order of VPC returned by AWS, which might be a bit surprising to the user seeing their cluster spreaded across multiple VPCs.

Comment (concretevitamin) [reply sky/skylet/providers/gcp/config.py]: It seems like the aws.py snippet quoted above will return 0 or 1 unique VPC identified by name / security group, so this scenario won't happen? Shall we talk more offline?

Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py]: Ahh, good point! Since we will always specify the security group name, the subnets got later should always be in the same VPC. In that case, it should be fine.

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:881]:
```suggestion
                "This is probably due to the region being disabled in the account/project_id."
```

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:817]: Nit: Doesn't feel very clean calling _list_subnets multiple times.

Is it possible to use the pattern in AWS's config.py? Call _list_subnets() once to get all `[(subnet, vpc, ...)]` and directly sort + filter this list.

Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py:815]: Good point! Just changed the implementation to list subnets first, which will save at most #VPC times gcp API call.

Tested:
- [x] `sky launch --cpus 2 --cloud gcp`
- [x] `sky launch --gpus L4 --cloud gcp` (default VPC does not have the first region to be tried for L4 GPU, i.e. us-east-4, and it will correctly use skypilot-vpc instead)

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:789]: Good catch! Q: it seems like sky/provision/aws/config.py doesn't have such back-compat code?

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:840]: nit: Probably worth writing
```
subnet_to_vpc_name = lambda s: s["network"].split("/")[-1]
```
and replace L840, L853?

**Review by concretevitamin — APPROVED** [review]
LGTM @Michaelvll

File: sky/skylet/providers/gcp/config.py
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py]: Modified. Thanks!

File: sky/skylet/providers/gcp/config.py
Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py:794]: Good question! I just tried to create a new AWS VPC and after a cluster is provisioned, and run the `sky launch` again on that cluster. It seems `config.py` choose a different VPC, but the launching of that cluster still succeeded and the VPC of that cluster is still the old one. Probably, AWS's API ignores the different VPC for an existing VM.

Comment (concretevitamin) [reply sky/skylet/providers/gcp/config.py:794]: Got it. Maybe worth a test/comment here on whether GCP has the same property.

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [inline sky/skylet/providers/gcp/config.py:802]:
```suggestion
    # the cluster, as GCP will ignore them.
```

File: sky/provision/aws/config.py
Comment (concretevitamin) [inline sky/provision/aws/config.py:219]:
```suggestion
    # the cluster, as AWS will ignore them.
```

File: sky/skylet/providers/gcp/config.py
Comment (concretevitamin) [reply sky/skylet/providers/gcp/config.py]: That sounds good. 

Q: before step 3, is the single-node cluster stopped, or is it live? Wondering if the behavior applies to relaunching(starting) a stopped VM as wel.

Comment (Michaelvll) [reply sky/skylet/providers/gcp/config.py]: I tried both stopped and UP clusters, and it seems the steps above work correctly for both.

**Pull request merged**
